File: cses/07_mathematics/1723_graph-paths-i/solution.py
# Graph Paths I - https://cses.fi/problemset/task/1723
# Đếm số walk (đường đi có thể lặp cạnh) độ dài đúng k từ đỉnh 1 đến đỉnh n,
# kết quả modulo 1e9+7. Dùng matrix exponentiation của ma trận kề.
# Tối ưu bằng kỹ thuật "bignum packing": mỗi hàng ma trận được đóng gói thành
# một số nguyên lớn, mỗi phần tử chiếm SHIFT bit -> nhân ma trận nhanh hơn.

import sys


def solve() -> None:
    data = sys.stdin.buffer.read().split()
    idx = 0
    n = int(data[idx]); idx += 1
    m = int(data[idx]); idx += 1
    k = int(data[idx]); idx += 1

    MOD = 1_000_000_007

    # SHIFT = 67: mỗi ô (slot) rộng 67 bit.
    # Giá trị lớn nhất của một ô = n*(MOD-1)^2 <= 100*(1e9+6)^2 ~ 1.0e20 < 2^67.
    SHIFT = 67
    MASK = (1 << SHIFT) - 1

    # Ma trận kề: A[i][j] = số cạnh từ i sang j.
    A = [[0] * n for _ in range(n)]
    for _ in range(m):
        a = int(data[idx]); idx += 1
        b = int(data[idx]); idx += 1
        A[a - 1][b - 1] += 1

    def mat_mult(X, Y):
        # Tính X*Y (mod MOD). Đóng gói các hàng của Y thành số nguyên lớn.
        Yenc = []
        for row in Y:
            e = 0
            for j in range(n - 1, -1, -1):
                e = (e << SHIFT) | row[j]
            Yenc.append(e)
        Z = []
        for i in range(n):
            Xi = X[i]
            acc = 0
            for kk in range(n):
                a = Xi[kk]
                if a:
                    acc += a * Yenc[kk]
            zi = [0] * n
            for j in range(n):
                zi[j] = (acc >> (j * SHIFT)) & MASK
                if zi[j] >= MOD:
                    zi[j] %= MOD
            Z.append(zi)
        return Z

    # Binary matrix exponentiation: result = A^k.
    # result khởi tạo = ma trận đơn vị.
    result = [[1 if i == j else 0 for j in range(n)] for i in range(n)]
    base = A
    e = k
    while e:
        if e & 1:
            result = mat_mult(result, base)
        e >>= 1
        if e:
            base = mat_mult(base, base)

    ans = result[0][n - 1] % MOD
    sys.stdout.write(str(ans) + "\n")


solve()

File: cses/07_mathematics/1723_graph-paths-i/solution.cpp
#include <iostream>
#include <vector>

using namespace std;

namespace {
constexpr long long MOD = 1'000'000'007LL;
using Matrix = vector<vector<long long>>;

// Nhân hai ma trận n x n theo modulo MOD (dùng cho phép bình phương base).
// Bỏ qua các phần tử bằng 0 để tăng tốc.
Matrix multiply(const Matrix& left, const Matrix& right) {
    const size_t n = left.size();
    Matrix product(n, vector<long long>(n));

    for (size_t i = 0; i < n; ++i) {
        for (size_t middle = 0; middle < n; ++middle) {
            if (left[i][middle] == 0) {
                continue;
            }
            const long long coefficient = left[i][middle];
            for (size_t j = 0; j < n; ++j) {
                product[i][j] = (product[i][j] + coefficient * right[middle][j]) % MOD;
            }
        }
    }
    return product;
}

// Nhân vector hàng với ma trận (row * matrix) theo modulo MOD.
// Chỉ cần cập nhật hàng trạng thái thay vì cả ma trận A^k, nhanh hơn.
vector<long long> multiply(const vector<long long>& row, const Matrix& matrix) {
    const size_t n = row.size();
    vector<long long> product(n);
    for (size_t i = 0; i < n; ++i) {
        if (row[i] == 0) {
            continue;
        }
        for (size_t j = 0; j < n; ++j) {
            product[j] = (product[j] + row[i] * matrix[i][j]) % MOD;
        }
    }
    return product;
}
}  // namespace

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    size_t n, m;
    long long k;
    cin >> n >> m >> k;

    // Xây ma trận kề: base[i][j] = số cạnh có hướng đi từ i sang j.
    Matrix base(n, vector<long long>(n));
    for (size_t edge = 0; edge < m; ++edge) {
        size_t from, to;
        cin >> from >> to;
        --from;
        --to;
        base[from][to] = (base[from][to] + 1) % MOD;
    }

    // paths là vector hàng, khởi tạo tại đỉnh 1 (index 0).
    // Nhân dần với A^k bằng binary exponentiation: sau vòng lặp paths = e_0 * A^k,
    // nên paths[n-1] chính là (A^k)[0][n-1] = số walk độ dài k từ đỉnh 1 tới đỉnh n.
    vector<long long> paths(n);
    paths[0] = 1;
    while (k > 0) {
        if (k & 1LL) {
            paths = multiply(paths, base);
        }
        k >>= 1LL;
        if (k > 0) {
            base = multiply(base, base);
        }
    }

    cout << paths[n - 1] << '\n';
    return 0;
}
